client.go

```diff
@@ -0,0 +1,297 @@
+	}
+	return nil
+}
+
+// Attempt to send metric to graphite server via connection
+func (c *Client) tryToSendToGraphite(metric string, conn net.Conn) error {
+	// If at any point "HOSTNAME" was used instead of real hostname - replace it
+	metric = strings.Replace(metric, "HOSTNAME", c.Lc.hostname, -1)
+
+	_, err := conn.Write([]byte(metric + "\n"))
+	if err != nil {
+		c.Lc.lg.Println("Write to server failed:", err.Error())
+		return err
+	}
+	backend := conn.RemoteAddr().String()
+	c.Mon.Increase(&c.Mon.stat[backend].sent, 1)
+	return nil
+}
+
+// Run go routine per carbon server to:
+//  1) Send data from retryFile to a carbon
+//  2) Send metrics from monitoring channel to a carbon
+//  3) Send metrics from the main channel to carbon
+//
+// And save everything to the retryFile on any error
+func (c Client) runBackend(backend string) {
+	retFile := path.Join(c.Conf.RetryDir, backend)
+	chanLock.Lock()
+	monChannel := c.monChannels[backend]
+	mainChannel := c.mainChannels[backend]
+	chanLock.Unlock()
+	// TODO: think about graceful shutdown with flush channels
+
+	for ; ; time.Sleep(time.Duration(c.Conf.ClientSendInterval) * time.Second) {
+		var connectionFailed bool
+
+		// Try to dial to Graphite server. If ClientSendInterval is 10 seconds - dial should be no longer than 1 second
+		conn, err := net.DialTimeout("tcp", backend, time.Duration(c.Conf.ConnectTimeout)*time.Second)
+		if err != nil {
+			c.Lc.lg.Println("Can not connect to graphite server: ", err.Error())
+			c.saveChannelToRetry(monChannel, len(monChannel), backend)
+			c.saveChannelToRetry(mainChannel, len(mainChannel), backend)
+			c.removeOldDataFromRetryFile(backend)
+			continue
+		}
+
+		// We set dead line for connection to write. It should be the rest of we have for client interval
+		err = conn.SetWriteDeadline(time.Now().Add(time.Duration(c.Conf.ClientSendInterval-c.Conf.ConnectTimeout-1) * time.Second))
+		if err != nil {
+			c.Lc.lg.Println("Can not set deadline for connection: ", err.Error())
+			connectionFailed = true
+		}
+
+		// We send retry file first, we have a risk to lose old data
+		// Metrics from retry file are counted as extra metrics per second to have a chance to send them
+		// Otherwise we would only save new incomming metrics and continuously lose part of buffer
+		if !connectionFailed {
+			retryFileMetrics, _ := readMetricsFromFile(retFile)
+			for numOfMetricFromFile, metricFromFile := range retryFileMetrics {
+				if numOfMetricFromFile < c.Lc.mainBufferSize {
+					err = c.tryToSendToGraphite(metricFromFile, conn)
+					if err != nil {
+						c.Lc.lg.Printf("Error happened in the middle of writing retry metrics. Resaving %d metrics\n", len(retryFileMetrics[numOfMetricFromFile:]))
+						// If we failed to write a metric to graphite - something is wrong with connection
+						c.saveSliceToRetry(retryFileMetrics[numOfMetricFromFile:], backend)
+						connectionFailed = true
+						break
+					} else {
+						c.Mon.Increase(&c.Mon.got.retry, 1)
```

Comment:
This is an error, it should be `c.Mon.stat[backend].retry`. Doesn't make a lot of sense, TBH.  
Maybe, we could rename it to `from_retry` and then put to the `stat` map